src/tools/clippy/clippy_lints/src/mixed_read_write_in_expression.rs

```diff
@@ -134,15 +136,56 @@ impl<'a, 'tcx> DivergenceVisitor<'a, 'tcx> {
     }
 }
 
+fn expr_might_diverge(e: &Expr<'_>) -> bool {
```

Comment:
We have a really elaborate implementation fo diverging expressions in `clippy_lints/src/manual_let_else.rs` called `expr_diverges`. This can probably also be used here. For this, moving that function to `clippy_utils` is probably the best thing to do.